documentation/FIR-MLIR-dialect.md

```diff
@@ -0,0 +1,204 @@
+# Investigating FIR as an MLIR Dialect
+
+## Introduction
+
+With the recent availability of the extensible MLIR framework as open-source, we've been investigating the potential to use MLIR as a substrate upon which to build a FIR dialect. This document will explain the motivations and the shape of this subproject using MLIR to define and implement a high-level operational representation of Fortran. (See also _Design: Fortran IR_ at [https://github.com/flang-compiler/f18/blob/master/documentation/FortranIR.md](https://github.com/flang-compiler/f18/blob/master/documentation/FortranIR.md) for more information.)
+
+### What is MLIR?
+
+The MLIR project is a recently open-sourced project on Github. MLIR (see [https://github.com/tensorflow/mlir](https://github.com/tensorflow/mlir) for more information on MLIR) is being built as a common multi-level intermediate representation (IR) for the TensorFlow machine learning tools. It provides a common IR framework for building multiple layers of representation called dialects. These dialects can be transformed (converted) from one to another allowing for optimizations and successive lowering of the IR down to backend code generators. One of these code generators is LLVM IR. MLIR is explicitly intended to be extensible such that other projects can add their own dialects.
+
```

Comment:
Will MLIR also be the pathway to non-LLVM legacy proprietary code generation?

Reply:
I do not know the answer to that.

Reply:
Can MLIR also be the pathway to non-LLVM legacy proprietary code generation?

Should MLIR also be the pathway to non-LLVM legacy proprietary code generation?

Reply:
> Can MLIR also be the pathway to non-LLVM legacy proprietary code generation?

Yes. MLIR is currently being used to target multiple code gens.

> Should MLIR also be the pathway to non-LLVM legacy proprietary code generation?

Not trying to be dodgy, but I haven't spent any time speculating on requirements, trade-offs, or assessing complications. Sorry.